### PR TITLE
AIP-84: Add UI batch recent dag runs endpoint

### LINKED ISSUE #42933: AIP-84 Add UI batch recent_dag_runs endpoint
In the new Dags list view we show a chart of recent dag runs:

Image

We should add a new endpoint `ui/dags/recent_dag_runs` to return an array of recent dag run arrays like so:
```
[{ dag_id: 'string', dag_runs: []}]
```

It should accept the same filters and pagination as the public get dags endpoint so that we can match them together.

This replaces the viewspy `/last_dagruns`.

### PR DESCRIPTION
**Pull request by jason810496**
closes: #42933
Same filter with public [`get_dags`](https://github.com/apache/airflow/blob/57500b6608fcbd3bfaa1ddbe6364899ac3f8c251/airflow/api_fastapi/core_api/routes/public/dags.py#L60-L68) endpoint, but remove order_by params.

### CODE REVIEW

File: airflow/api_fastapi/core_api/routes/ui/dags.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+    return RecentDAGRunsCollectionResponse(
+        total_dag_runs=total_entries,
+        total_dag_ids=len(dag_runs_by_dag_id),
+        recent_dag_runs=[
+            RecentDAGRunsResponse(
+                dag_id=dag_id,
+                dag_runs=[
+                    RecentDAGRun(
+                        start_date=dag_run.start_date,
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/ui/dags.py]: I think we should try to keep public and private serializers the same unless we specifically need a new representation of the object. (This will avoid drifts and make things easier to integrate to the public API).

Also in the response I think DagRuns should be grouped per Dag. For instance, you can subclass the public `DAGResponse` and add a `latest_dag_runs` field to it. And then reuse the `DAGCollectionResponse` from there. And same you can re-use the public `DagRunResponse` to serialize the DagRun.

Basically the UI endpoint returns exactly the same but with an additional `latest_dag_runs` field for each DAGs.

**Review by pierrejeambrun — COMMENTED** [review]
Nice

File: airflow/api_fastapi/core_api/routes/ui/dags.py (hunk 1 above)
Comment (jason810496) [reply airflow/api_fastapi/core_api/routes/ui/dags.py]: Sure, I will refactor this UI endpoint using the public serializers.

**Comment by jason810496** [comment]
### Refactor:
Used `DAGRunResponse`, `DAGCollectionResponse`, and `DAGResponse` from the public serializers for the UI batch of recent runs. However, the subclasses that inherit from the public serializers should also inherit from `OptionalModel`, as most of the fields should be optional.

File: airflow/api_fastapi/core_api/serializers/ui/dags.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+from airflow.api_fastapi.core_api.serializers.dag_run import DAGRunResponse
+from airflow.api_fastapi.core_api.serializers.dags import DAGCollectionResponse, DAGResponse
+from airflow.api_fastapi.core_api.serializers.optional import OptionalModel
+
+
+class RecentDAGRunResponse(DAGRunResponse, OptionalModel):
+    """Recent DAG Run response serializer."""
+
+    execution_date: datetime
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/serializers/ui/dags.py]: Execution date is deprecated I believe. Now this is the `logical` date. This particular one shouldn't have more 'optional' field than the base `DaGRunResopnse`. (nullable fields are already defined there if possible)

**Review by pierrejeambrun — COMMENTED** [review]
I am not sure I understand the `OptionalModel`. Why do those DAGRuns and DAGs omit fields in the private response compare to the public one ?

I would assume that **all** of the public fields are also defined and here in the private endpoint therefore removing the need for that `OptionalModel` class. Did you encounter a specific issue ?

**Comment by jason810496** [comment]
I think for this UI, the endpoint doesn't need to respond with so much information (compared with `/last_dagruns` in `views.py`). In `/last_dagruns`, it only needs `dag_id`, `start_date`, `end_date`, `state`, `execution_date`, `data_interval_start`, and `data_interval_end`.

Instead of marking the fields that aren't needed in this endpoint as optional in `DAGRunResponse`, I think inheriting from `OptionalModel` to mark the entire `RecentDAGRunResponse` would be clearer.

**Comment by jason810496** [comment]
**Resolved:**

Made this UI endpoint more consistent with the public endpoint by removing `OptionalModel` and using `DAGRunResponse` directly from the public endpoint instead.

**Review by pierrejeambrun — APPROVED** [review]
NIce

File: airflow/api_fastapi/core_api/serializers/ui/dags.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+from __future__ import annotations
+
+from pydantic import BaseModel
+
+from airflow.api_fastapi.core_api.serializers.dag_run import DAGRunResponse
+from airflow.api_fastapi.core_api.serializers.dags import DAGResponse
+
+
+class DAGWithLatestDagRunsResponse(DAGResponse):
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/serializers/ui/dags.py:26]: Now a `DAGWithLatestDagRunsResponse` is a `DAGResponse`. Allow to retrieve all the same properties with just an additional one `latest_dag_runs`. Makes it possible to remove the `type: ignore` in the collection.

Comment (pierrejeambrun) [reply airflow/api_fastapi/core_api/serializers/ui/dags.py:26]: Making it easie to convert one to another in the frontend and also really simple to move this to the public API once we think it's ready to.

File: airflow/api_fastapi/core_api/routes/ui/dags.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    offset: QueryOffset,
+    tags: QueryTagsFilter,
+    owners: QueryOwnersFilter,
+    dag_id_pattern: QueryDagIdPatternSearch,
+    dag_display_name_pattern: QueryDagDisplayNamePatternSearch,
+    only_active: QueryOnlyActiveFilter,
+    paused: QueryPausedFilter,
+    last_dag_run_state: QueryLastDagRunStateFilter,
+    session: Annotated[Session, Depends(get_session)],
+    dag_runs_limit: int = 10,
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/ui/dags.py:64]: This limit shouldn't the same as the `limit`. So we can specify `x` dags, but `y` dag runs per dag. I updated it accordingly.

**Comment by jason810496** [comment]
Hi @pierrejeambrun, the changes look good to me, but I have another question:

When should I utilize a parameter class (the classes declared in `common/parameters.py` with `to_orm` and `depends` methods) instead of adding query parameters directly (like `dag_runs_limit: int = 10` in this case)?

For example, in the Event Log endpoints, the query should leverage `paginated_select`, but the `paginated_select` function depends on a parameter class rather than query parameters. In this case, should I declare all filter classes in `common/parameters.py`, or would it be better to add a `FilterParam` with `dynamic_depends`, similar to `SortParam`, that could be used across all endpoints?

**Comment by pierrejeambrun** [comment]
>When should I utilize a parameter class (the classes declared in common/parameters.py with to_orm and depends methods) instead of adding query parameters directly (like dag_runs_limit: int = 10 in this case)?
For example, in the Event Log endpoints, the query should leverage paginated_select, but the paginated_select function depends on a parameter class rather than query parameters. In this case, should I declare all filter classes in common/parameters.py, or would it be better to add a FilterParam with dynamic_depends, similar to SortParam, that could be used across all endpoints?


The idea is if the parameter (query or path) is meant to be re-used in other endpoints and implement a common db operation (filtering, sorting, matching, anything), then it should go into `parameters` with our common parameters.

If like in this case it is a very specific one, in terms of parameters (a secondary limit index operating on nested ressources) in terms of naming "dag_runs_limit", and in terms of implementation (`.where(recent_runs_subquery.c.rank <= dag_runs_limit)` which is really specific to the subquery), then it's ok to leave it in the router and not bother creating a reusable 'parameters' out of if.
